File: experiments/source/cpp/affine_clamp.cpp
#include <loom/loom.hpp>

LOOM_ENTRY loom::Tensor1 affine_clamp(float scale, float bias, float lo, float hi, loom::Tensor1 x) {
  auto clamp = [=](float xi) -> float {
    auto value = (scale * xi) + bias;
    return value < lo ? lo : value > hi ? hi : value;
  };
  return loom::Tensor1::map(clamp, x);
}

File: experiments/source/python/affine_clamp.py
import loom


@loom.entry
def affine_clamp(scale: float, bias: float, lo: float, hi: float, x: loom.Tensor1) -> loom.Tensor1:
    def clamp(xi: float) -> float:
        value = (scale * xi) + bias
        return lo if value < lo else hi if value > hi else value

    return loom.Tensor1.map(clamp, x)
